### PR TITLE
Python: have decorators use docstring, preserve metadata, cleanup and add more docs

### PR DESCRIPTION
**Pull request by jjhenkel**
### Motivation and Context

Taking in some feedback on decorators, cleaning them up, making sure they preserve the original function's metadata, adding more docs. For `sk_function` if there is no explicit description, the function's `docstring` will be used.

### Description

I wanted to take another pass at decorators based off some of the original ideas in @dluc's v0 and feedback we've been receiving. This PR does the following:

1. An `sk_function` now will take both the name of the decorated function (as the Native Function's name) and the docstring of the decorated function (as the Native Function's description) if no explicit name/description is set.
2. The `sk_function` and `sk_function_context_parameter` decorators use `@wraps` to preserve the original function's metadata.
3. The `sk_function` decorator will remove the `_async` suffix, if present, and no explicit name is given. (This seems natural, in the Prompt Template it feels a little weird to write `my_function_async` since we don't really have a concept of sync/async in the Prompt Template language.) 
4. Added some more thorough docs, some comments, made typechecker in VSCode happier (w/ `setattr`)

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by jjhenkel** [comment]
Very open to feedback on this! @JTremb I know you did the original decorators fixup, so I'd especially love your feedback! Also @dluc for thoughts (since you had the idea to use docstring originally in your v0).

File: python/semantic_kernel/skill_definition/sk_function_decorator.py [hunk 1]
```diff
@@ -1,29 +1,74 @@
...
+            Removes a trailing _async from the function name.
+            """
+            if name.endswith("_async"):
+                return name[:-6]
+            return name
+
+        # Switched to setattr to make the type checker happy.
+        setattr(wrapper, "__sk_function__", True)
+        setattr(wrapper, "__sk_function_name__", name or _cleanup(func.__name__))
+        setattr(wrapper, "__sk_function_description__", description or func.__doc__)
```
Comment (JTremb) [inline python/semantic_kernel/skill_definition/sk_function_decorator.py:69]: Like the option of giving an explicit description to the decorator or else using the docstring.
For the docstring, should we parse to only keep the `usage` section or keep the whole thing as the description?

**Comment by dluc** [comment]
@jjhenkel I think this would be great, could we go through all the details at our next sync? I think we need to think about documentation and examples. I'd like to know more about special cases, e.g. new lines, and what happens if the description contains more than just the description, which seems to be a common practice in python

**Comment by dluc** [comment]
a couple of conflicts to address

**Comment by JTremb** [comment]
For the docstring is there a preferred format to use ? e.g. to generate the documentation with Sphynx (unless I missed it I don't I saw it)

File: python/semantic_kernel/skill_definition/sk_function_decorator.py (hunk 1 above)
Comment (jjhenkel) [reply python/semantic_kernel/skill_definition/sk_function_decorator.py:69]: This is a really good question... I think we need to understand how the planner uses the description better before giving a precise answer. In general, I think throwing the whole docstring in is workable, but probably not optimal. (Probably want to do some post processing, like what you mentioned)

Agree w/ @dluc on chatting about it! I'll make sure you have the invite to chat too so we can get everyone's perspective!

**Comment by jjhenkel** [comment]
@dluc many tests added for defining how we pre-process the docstring to use as a description. (Per your suggestion, we take everything up to first blank line, normalize whitespace.)

**Comment by shawncal** [comment]
FYI: Python PRs on hold while we stage to merge /python into main. See https://github.com/microsoft/semantic-kernel/pull/423

**Comment by shawncal** [comment]
@jjhenkel Now that Python is in main, I tried retargeting this PR to merge into main. As soon as I did that, it immediately closed the PR, and now I cannot reopen, nor can I retarget back to Python. My apologies. Go ahead and resubmit the PR to target main, if it's still valid.

**Pull request closed without merging**